PHPCompatibility/Sniffs/Syntax/InterpolatedStringDereferencingSniff.php

```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * PHPCompatibility, an external standard for PHP_CodeSniffer.
+ *
+ * @package   PHPCompatibility
+ * @copyright 2009-2020 PHPCompatibility Contributors
+ * @license   https://opensource.org/licenses/LGPL-3.0 LGPL3
+ * @link      https://github.com/PHPCompatibility/PHPCompatibility
+ */
+
+namespace PHPCompatibility\Sniffs\Syntax;
+
+use PHPCompatibility\Sniff;
+use PHPCSUtils\Utils\TextStrings;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Util\Tokens;
+
+/**
+ * Detect dereferencing of interpolated strings.
+ *
+ * PHP version 8.0
+ *
+ * @link https://wiki.php.net/rfc/variable_syntax_tweaks#interpolated_and_non-interpolated_strings
+ * @link https://www.php.net/manual/en/language.types.string.php#language.types.string.parsing
+ *
+ * @since 10.0
+ */
+class InterpolatedStringDereferencingSniff extends Sniff
+{
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @since 10.0
+     *
+     * @return array
+     */
+    public function register()
+    {
+        return [
+            \T_DOUBLE_QUOTED_STRING,
+            \T_HEREDOC,
+            \T_NOWDOC,
+        ];
+    }
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @since 10.0
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token
+     *                                               in the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        // PHP 8.0 supports dereferencing interpolated strings
+        if ($this->supportsAbove('8.0')) {
+            return;
+        }
+
+        $tokens = $phpcsFile->getTokens();
+
+        // String doesn't use interpolation
+        $string = $tokens[$stackPtr]['content'];
+        if ($this->stripVariables($string) === $string) {
+            return;
+        }
+
+        // String is not being dereferenced
+        $skipTokens       = Tokens::$emptyTokens + [\T_END_HEREDOC, \T_END_NOWDOC];
+        $nextNonEmpty     = $phpcsFile->findNext($skipTokens, ($stackPtr + 1), null, true);
+        $nextNonEmptyCode = $tokens[$nextNonEmpty]['code'];
+        if ($nextNonEmptyCode !== \T_OPEN_SHORT_ARRAY
+            && $nextNonEmptyCode !== \T_OBJECT_OPERATOR
+        ) {
+            return;
+        }
+
+        $error   = 'Dereferencing of interpolated strings is not supported before PHP 8.0';
+        $isError = true;
+        $this->addMessage($phpcsFile, $error, $stackPtr, $isError);
```

Comment:
```suggestion
        $phpcsFile->addError(
            'Dereferencing of interpolated strings was not supported in PHP 7.4 or earlier',
            $stackPtr,
            'Found'
        );
```

The `Sniff::addMessage()` method is a convenience wrapper for sniffs which depending on the `testVersion` may throw an error or a warning.

As this sniff will only ever throw an error, there is no need to use the wrapper method and the PHPCS native `addError()` method can be used directly.

Also:
* No need for the "only used once" `$error` variable.
* While the default error code is `Found`, we tend to be explicit about the error code.
* In the error message itself, we tend to use phrasing like "_not supported in PHP 7.4 or earlier_" to prevent confusing people.
    Also see: #934